### PR TITLE
Implement WAL static analysis to ensure that new entries will be properly annotated

### PR DESCRIPTION
**Pull request by serathius**
Part of https://github.com/etcd-io/etcd/issues/13168 

cc @hexfusion @ptabor

### CODE REVIEW

File: server/storage/wal/version.go [hunk 1]
```diff
@@ -53,32 +53,69 @@ func (w *walVersion) MinimalEtcdVersion() *semver.Version {
...
 	return maxVer
 }
 
-func etcdVersionFromEntry(ent raftpb.Entry) *semver.Version {
-	msgVer := etcdVersionFromMessage(proto.MessageReflect(&ent))
-	dataVer := etcdVersionFromData(ent.Type, ent.Data)
-	return maxVersion(msgVer, dataVer)
+type Visitor func(path protoreflect.FullName, ver *semver.Version) error
+
+func VisitFileDescriptor(file protoreflect.FileDescriptor, visitor Visitor) error {
```
Comment (ptabor) [inline server/storage/wal/version.go:69]: Please document that this method calls visitor on each field and enum value with the 
version coming from the configured option or nil if unset.

File: server/storage/wal/version.go [hunk 2]
```diff
@@ -98,43 +135,106 @@ func etcdVersionFromData(entryType raftpb.EntryType, data []byte) *semver.Versio
...
+		return err
+	}
+	return visitEnumValue(enum.Values().Get(int(number)), visitor)
+}
+
+func visitEnumValue(enum protoreflect.EnumValueDescriptor, visitor Visitor) error {
+	valueOpts := enum.Options().(*descriptorpb.EnumValueOptions)
 	if valueOpts != nil {
-		maxVer = maxVersion(maxVer, etcdVersionFromOptionsString(valueOpts.String()))
+		ver, _ := etcdVersionFromOptionsString(valueOpts.String())
```
Comment (ptabor) [inline server/storage/wal/version.go:219]: I wonder whether we are not parsing these options too frequently (e.g. on each interpreted _data_ enum). 
Alternatively we could build a map(s):
  -  field-> version
  - enum -> version 
The way we build the version file... and during the data processing just keep performing lookups in the map.

File: tools/proto-annotations/main.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	if len(errs) != 0 {
+		for _, err := range errs {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func handleEtcdVersion() (errs []error) {
```
[the remark this answers is not in the document]
Comment (serathius) [reply tools/proto-annotations/main.go:58]: Done

File: server/storage/wal/version.go (hunk 1 above)
Comment (serathius) [reply server/storage/wal/version.go:69]: Done

File: server/storage/wal/version.go (hunk 2 above)
Comment (serathius) [reply server/storage/wal/version.go:222]: Sure, we will definitely need this for version API calls as they will be latency sensitive, however I would skip this optimization of static analysis.

File: scripts/update_proto_annotations.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
```
Comment (ptabor) [inline scripts/update_proto_annotations.sh:2]: Please add a comment what it does and when to use it.

File: tools/proto-annotations/cmd/etcd_version.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	return errs
+}
+
+func allEtcdVersionAnnotations() (annotations []etcdVersionAnnotation, err error) {
+	var fileAnnotations []etcdVersionAnnotation
+	protoregistry.GlobalFiles.RangeFiles(func(file protoreflect.FileDescriptor) bool {
+		switch string(file.Package()) {
+		// Skip external packages that are not expected to have etcd version annotation.
+		case "io.prometheus.client", "grpc.binarylog.v1", "google.protobuf", "google.rpc", "google.api":
```
Comment (ptabor) [inline tools/proto-annotations/cmd/etcd_version.go:64]: nit: please move it to the top level constant.

File: server/storage/wal/version.go (hunk 2 above)
Comment (ptabor) [reply server/storage/wal/version.go:222]: Aren't we afraid that  start time of etcd server that will keep processing WAL log (ReadWALVersion) will get significantly slower. Have you tested it with realistic: O(64MB) Wal files ?

**Review by ptabor — APPROVED** [review]

File: server/storage/wal/version.go (hunk 2 above)
Comment (serathius) [reply server/storage/wal/version.go:222]: We don't verify WAL during server start, we do it during downgrade and after we created a snapshot.

File: scripts/update_proto_annotations.sh (hunk 4 above)
Comment (serathius) [reply scripts/update_proto_annotations.sh:2]: Done

File: tools/proto-annotations/cmd/etcd_version.go (hunk 5 above)
Comment (serathius) [reply tools/proto-annotations/cmd/etcd_version.go:64]: Done

**Comment by serathius** [comment]
@ptabor  PTAL

**Comment by ptabor** [comment]
Thank you !

**Pull request merged**